File: day-2/python/main.py
"""Python implementation for AoC task."""

import itertools
import typing

from interpreter import Interpreter

TO_FIND = 19690720


def get_codes() -> typing.List[int]:
    """Fetch intcode from the input file."""
    with open("../input") as f:
        codes = [int(x) for x in f.read().split(",")]

    return codes


if __name__ == "__main__":
    codes = get_codes()

    # Part 1

    # Account for error
    codes[1] = 12
    codes[2] = 2

    inter = Interpreter(codes[:])

    while not inter.halted:
        inter.step()

    print(f"Part 1: {inter.codes[0]}")

    # Part 2

    for noun, verb in itertools.product(range(0, 100), repeat=2):
        codes[1] = noun
        codes[2] = verb

        inter.codes = codes[:]
        inter.reset()

        while not inter.halted:
            inter.step()

        if inter.codes[0] == TO_FIND:
            print(f"Part 2: 100 * {noun} + {verb} = {100 * noun + verb}")
            break

File: day-2/rust/src/main.rs
use std::fs::File;
use std::io::Read;

mod interpreter;

const TO_FIND: i64 = 19690720;

fn main() {
    let input = get_input().expect("Could not open input, does the file exist?");

    let mut codes = input.split_terminator(",")
        .map(|x| x.trim())
        .map(|x| x.parse::<i64>().unwrap())
        .collect::<Vec<i64>>();

    // Account for Error
    codes[1] = 12;
    codes[2] = 2;

    let mut interpreter = interpreter::Interpreter::new(codes);

    while interpreter.is_running {
        interpreter.step();
    }

    println!("Part 1: {}", interpreter.fetch(0));

    'outer: for i in 0..=99 {
        'inner: for j in 0..=99 {
            let mut codes = input.split_terminator(",")
                .map(|x| x.trim())
                .map(|x| x.parse::<i64>().unwrap())
                .collect::<Vec<i64>>();

            codes[1] = i;
            codes[2] = j;
            let mut interpreter = interpreter::Interpreter::new(codes);

            while interpreter.is_running {
                interpreter.step();
            }

            if interpreter.fetch(0) == TO_FIND {
                println!("Part 2: 100 * {} + {} = {}", i, j, 100 * i + j);
                break 'outer;
            }
        }
    }
}

fn get_input() -> Result<String, std::io::Error> {
    let mut f = File::open("../input")?;

    let mut buf = String::new();
    f.read_to_string(&mut buf)?;

    Ok(buf)
}
